From 6dc8b3d862f8991486d18e4a3e02fdd33cd3b8c0 Mon Sep 17 00:00:00 2001
From: David Reaver <me@davidreaver.com>
Date: Sun, 19 Jan 2025 16:45:48 -0800
Subject: [PATCH 5/5] samples: Implement removing directories in sample_kernfs

Signed-off-by: David Reaver <me@davidreaver.com>
---
 samples/kernfs/sample_kernfs.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/samples/kernfs/sample_kernfs.c b/samples/kernfs/sample_kernfs.c
index 70640365cc3f..8ba8d18427c3 100644
--- a/samples/kernfs/sample_kernfs.c
+++ b/samples/kernfs/sample_kernfs.c
@@ -79,8 +79,25 @@ static int sample_kernfs_mkdir(struct kernfs_node *parent_kn, const char *name,
 	return sample_kernfs_populate_dir(dir_kn);
 }
 
+static int sample_kernfs_rmdir(struct kernfs_node *kn)
+{
+  	// Free our sample_kernfs_directory struct, stored in the node's private
+  	// data.
+  	if (kn->priv)
+  		kfree(kn->priv);
+
+	// kernfs_remove_self safely removes the node while also breaking out of
+	// kernfs active protection (see kernfs_break_active_protection). This
+	// is necessary to avoid a deadlock because the kernfs function that
+	// called rmdir is holding the lock.
+	kernfs_remove_self(kn);
+
+  	return 0;
+}
+
 static struct kernfs_syscall_ops sample_kernfs_kf_syscall_ops = {
 	.mkdir		= sample_kernfs_mkdir,
+	.rmdir		= sample_kernfs_rmdir,
 };
 
 static int sample_kernfs_get_tree(struct fs_context *fc)
